### PR TITLE
Add option for capture the command output to script log

### LINKED ISSUE #8: [Feature] - Capture command output in script log
Hi

This script is a little gem 🥳

But I have a little feature request: In my scenario I am running HTTP web requests and I want to somehow give the user a hint if something went wrong (e.g. a condition on the server's side is not met)
(the server I am requesting this is also hosted by me, returning some json data)

For now it would be enough to store it to the script log. Is there a simple solution to achieve this?

### PR DESCRIPTION
**Pull request by IT-VBFK**
Proposal for #8 

Closes #8

- [x]  Testing with OBS (macOS)
- [x]  Testing with OBS (Windows)

### CODE REVIEW

**Comment by marklagendijk** [comment]
@IT-VBFK awesome that you are picking this up!
Does `io.popen` by any chance _not_ popup a window when it's executing the command (on Windows)?
That would be even more awesome, because that is the main issue this script currently has.

**Comment by IT-VBFK** [comment]
I don't know actually :D

WIll test this tomorrow (on windows and macOS)

**Comment by IT-VBFK** [comment]
Ok, checked on macOS now

https://recordit.co/trDFNsN7ku

this seems to work like excpected

**Comment by IT-VBFK** [comment]
io.popen opens a cmd windows on windows as well for a blink of an eye.

After googling a bit: this is only possible with external lua extensions, which are not built in in the lua obs uses.

File: scene_execute_command.lua [hunk 1]
```diff
@@ -51,6 +51,7 @@ end
 function script_properties()
 	local props = obs.obs_properties_create()
 
+	obs.obs_properties_add_bool(props, "capture_script_output", "Capture the script output in script log")
```
Comment (marklagendijk) [inline scene_execute_command.lua]: Consistency: use 'command' instead of 'script'

File: scene_execute_command.lua [hunk 2]
```diff
@@ -88,14 +89,30 @@ function handle_scene_change()
 	local scene = obs.obs_frontend_get_current_scene()
 	local scene_name = obs.obs_source_get_name(scene)
 	local scene_enabled = obs.obs_data_get_bool(settings, "scene_enabled_" .. scene_name)
+	local capture_script_output = obs.obs_data_get_bool(settings, "capture_script_output")
```
Comment (marklagendijk) [inline scene_execute_command.lua]: Consistency: use 'command' instead of 'script'

File: scene_execute_command.lua (hunk 1 above)
Comment (marklagendijk) [reply scene_execute_command.lua]: I would use the term 'log' instead of 'capture'.

File: scene_execute_command.lua [hunk 3]
```diff
@@ -88,14 +89,30 @@ function handle_scene_change()
...
+
 	if scene_enabled then
 		local command = obs.obs_data_get_string(settings, "command")
 		local scene_value = obs.obs_data_get_string(settings, "scene_value_" .. scene_name)
 		local scene_command = string.gsub(command, "SCENE_VALUE", scene_value)
 		obs.script_log(obs.LOG_INFO, "Activating " .. scene_name .. ". Executing command:\n  " .. scene_command)
-		os.execute(scene_command)
+
+		if capture_script_output then
```
Comment (marklagendijk) [inline scene_execute_command.lua]: I would move the whole if ... else block into the function `capture_output` and rename that function to `execute_command`

**Comment by marklagendijk** [comment]
@ IT-VBFK thanks a lot for this PR!
I did a code review and added a few (nitpicky) comments.

**Comment by IT-VBFK** [comment]
Ok thanks

What about using io.popen in every case, since it makes no difference?

This allows an other refacoring

**Comment by marklagendijk** [comment]
> What about using io.popen in every case, since it makes no difference?

Sure, go for it!

File: scene_execute_command.lua [hunk 4]
```diff
@@ -97,19 +97,18 @@ function handle_scene_change()
 		local scene_command = string.gsub(command, "SCENE_VALUE", scene_value)
 		obs.script_log(obs.LOG_INFO, "Activating " .. scene_name .. ". Executing command:\n  " .. scene_command)
 
+  local command_output = capture_output(scene_command)
+		
 		if log_command_output then
```
Comment (ITaluone) [inline scene_execute_command.lua]:
```suggestion
                local command_output = capture_output(scene_command)

		if log_command_output then
```

File: scene_execute_command.lua [hunk 5]
```diff
@@ -88,14 +89,29 @@ function handle_scene_change()
...
+		if log_command_output then
+			obs.script_log(obs.LOG_INFO, "Command output: " .. command_output)
+		end
 	else
 		obs.script_log(obs.LOG_INFO, "Activating " .. scene_name .. ". Command execution is disabled for this scene.")
 	end
 	obs.obs_source_release(scene);
 end
+
+function execute_output(command)
```
Comment (marklagendijk) [inline scene_execute_command.lua]: Name should be `execute_command`

File: scene_execute_command.lua [hunk 6]
```diff
@@ -88,14 +89,29 @@ function handle_scene_change()
...
+	local log_command_output = obs.obs_data_get_bool(settings, "log_command_output")
+
 	if scene_enabled then
 		local command = obs.obs_data_get_string(settings, "command")
 		local scene_value = obs.obs_data_get_string(settings, "scene_value_" .. scene_name)
 		local scene_command = string.gsub(command, "SCENE_VALUE", scene_value)
 		obs.script_log(obs.LOG_INFO, "Activating " .. scene_name .. ". Executing command:\n  " .. scene_command)
-		os.execute(scene_command)
+
+  local command_output = capture_output(scene_command)
```
Comment (marklagendijk) [inline scene_execute_command.lua]: Mismatch with function name

Comment (IT-VBFK) [reply scene_execute_command.lua]: On the phone right now ☺️
Sorry

**Comment by marklagendijk** [comment]
Thanks for the work and quick reactions!
I released a new version both on GitHub and the OBS versions.

**Pull request merged**